go/networkqualityd.go

```diff
@@ -116,44 +176,185 @@ func main() {
 				}
 			}
 		}
-		if announceResponder, announceHandle, announceErr := configureAnnouncer(ips, *configName, *configPort); announceErr == nil {
-			defer announceResponder.Remove(announceHandle)
-			go announceResponder.Respond(operatingCtx)
+	}
+
+	for port, scheme := range portScheme {
+		var hostPort string
+		if port == 80 || port == 443 {
+			hostPort = *publicName
 		} else {
-			log.Printf("Warning: Could not announce the server instance: %v.\n", announceErr)
+			hostPort = fmt.Sprintf("%s:%d", *publicName, port)
 		}
-	}
 
-	var wg sync.WaitGroup
-	wg.Add(1)
+		m := &Server{configName: *configName,
+			publicHostPort: hostPort,
+			template:       tmpl,
+			enableCORS:     *enableCORS,
+			contextPath:    *contextPath,
+			scheme:         scheme,
+		}
+
+		mux := http.NewServeMux()
+		mux.HandleFunc(m.contextPath+"/config", m.configHandler)
+		for pattern, handler := range BulkHandlers(m.contextPath, *enableCORS) {
+			mux.HandleFunc(pattern, handler)
+		}
+
+		var nl net.Listener
+		var err error
+
+		listenConfig := net.ListenConfig{
+			Control: func(network, address string, conn syscall.RawConn) error {
+				var setsockoptErr error
+				if *socketSendBuffer > 0 {
+					log.Printf("setting TCP_NOTSENT_LOWAT to %d", *socketSendBuffer)
+					if err := conn.Control(func(fd uintptr) {
+						setsockoptErr = syscall.SetsockoptInt(int(fd), unix.IPPROTO_TCP, unix.TCP_NOTSENT_LOWAT, int(*socketSendBuffer))
+					}); err != nil {
+						return err
+					}
+					if setsockoptErr != nil {
+						return setsockoptErr
+					}
+				}
 
-	server := &http.Server{}
-	server.Addr = fmt.Sprintf("%s:%d", *listenAddr, *configPort)
-	server.Handler = mux
+				if tos > 0 {
```

Comment:
Does it make sense to put this code in platform-specific file (the way that we have certain things in the client)?

Reply:
actually, I split this up. To be refined at a later date (for now I have a !windows implementation and windows implementation.j)